homeassistant/components/trafikverket_train/config_flow.py

```diff
@@ -0,0 +1,167 @@
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+import homeassistant.helpers.config_validation as cv
+import homeassistant.util.dt as dt_util
+
+from .const import CONF_FROM, CONF_TIME, CONF_TO, DOMAIN
+
+ERROR_INVALID_AUTH = "Source: Security, message: Invalid authentication"
+ERROR_INVALID_STATION = "Could not find a station with the specified name"
+ERROR_MULTIPLE_STATION = "Found multiple stations with the specified name"
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_API_KEY): cv.string,
+        vol.Required(CONF_FROM): cv.string,
+        vol.Required(CONF_TO): cv.string,
+        vol.Optional(CONF_TIME): cv.string,
+        vol.Required(CONF_WEEKDAY, default=WEEKDAYS): cv.multi_select(
+            {day: day for day in WEEKDAYS}
+        ),
+    }
+)
+DATA_SCHEMA_REAUTH = vol.Schema(
+    {
+        vol.Required(CONF_API_KEY): cv.string,
+    }
+)
+
+
+class TVTrainConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Trafikverket Train integration."""
+
+    VERSION = 1
+
+    entry: config_entries.ConfigEntry
+
+    async def validate_input(
+        self, api_key: str, train_from: str, train_to: str
+    ) -> None:
+        """Validate input from user input."""
+        web_session = async_get_clientsession(self.hass)
+        train_api = TrafikverketTrain(web_session, api_key)
+        await train_api.async_get_train_station(train_from)
+        await train_api.async_get_train_station(train_to)
+
+    async def async_step_reauth(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle re-authentication with Trafikverket."""
+
+        self.entry = self.hass.config_entries.async_get_entry(self.context["entry_id"])
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Confirm re-authentication with Trafikverket."""
+        errors: dict[str, str] = {}
+
+        if user_input:
+            api_key = user_input[CONF_API_KEY]
+
+            try:
+                await self.validate_input(
+                    api_key, self.entry.data[CONF_FROM], self.entry.data[CONF_TO]
+                )
+            except ValueError as err:
+                if str(err) == ERROR_INVALID_AUTH:
+                    errors["base"] = "invalid_auth"
+                elif str(err) == ERROR_INVALID_STATION:
+                    errors["base"] = "invalid_station"
+                elif str(err) == ERROR_MULTIPLE_STATION:
+                    errors["base"] = "more_stations"
+                else:
+                    errors["base"] = "cannot_connect"
+            else:
+                assert self.entry is not None
+
+                if (
+                    f"{self.entry.data[CONF_FROM]}-{self.entry.data[CONF_TO]}-{self.entry.data.get(CONF_TIME)}-{self.entry.data[CONF_WEEKDAY]}"
```

Comment:
Does this need to be normalized in some way?  What happens if you enter / copy&paste and some of the data in upper case, lower case, mixed case, with leading spaces, with null bytes, with a trailing new line, etc?

Reply:
You're right so I did set in for lowercase and removing any whitespace. Not sure if there is a better option?

Reply:
Maybe casefold?

Reply:
In use so this one should be fine then.